google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,359 @@
+    if (!this.BQTableSchema.equals(updatedSchema)) {
+      this.descriptor =
+          BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(updatedSchema);
+      this.BQTableSchema = updatedSchema;
+      streamWriter.refreshAppend();
+    }
+  }
+
+  /** Placeholder. Will update to include JsonToProtoMessage conversion. */
+  public synchronized ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    ApiFuture<AppendRowsResponse> appendResponseFuture = this.streamWriter.append(message);
+    ApiFutures.<AppendRowsResponse>addCallback(
+        appendResponseFuture,
+        new ApiFutureCallback<AppendRowsResponse>() {
+          @Override
+          public void onSuccess(AppendRowsResponse response) {
+            if (response.hasUpdatedSchema()) {
+              try {
+                updateDescriptor(response.getUpdatedSchema());
+              } catch (Exception e) {
+                // Deal with exceptions
+              }
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t) {}
+        });
+    return appendResponseFuture;
+  }
+
+  /**
+   * Gets streamName
+   *
+   * @return String
+   */
+  public String getStreamName() {
+    return this.streamName;
+  }
+
+  /**
+   * Gets current descriptor
+   *
+   * @return Descriptor
+   */
+  public Descriptor getDescriptor() {
+    return this.descriptor;
+  }
+
+  /**
+   * Gets current BQTableSchema
+   *
+   * @return Table.TableSchema
+   */
+  public Table.TableSchema getBQTableSchema() {
+    return this.BQTableSchema;
+  }
+
+  /** Sets all StreamWriter settings. */
+  private void setStreamWriterSettings(
+      StreamWriter.Builder builder,
+      TransportChannelProvider channelProvider,
+      CredentialsProvider credentialsProvider,
+      BatchingSettings batchingSettings,
+      RetrySettings retrySettings,
+      ExecutorProvider executorProvider,
+      String endpoint) {
+    if (channelProvider != null) {
+      builder.setChannelProvider(channelProvider);
+    }
+    if (credentialsProvider != null) {
+      builder.setCredentialsProvider(credentialsProvider);
+    }
+    if (batchingSettings != null) {
+      builder.setBatchingSettings(batchingSettings);
+    }
+    if (retrySettings != null) {
+      builder.setRetrySettings(retrySettings);
+    }
+    if (executorProvider != null) {
+      builder.setExecutorProvider(executorProvider);
+    }
+    if (endpoint != null) {
+      builder.setEndpoint(endpoint);
+    }
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder with BigQuery client being initialized by
+   * StreamWriter by default.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param BQTableSchema schema used to convert Json to proto messages.
+   * @return Builder
+   */
+  public static Builder newBuilder(String streamName, Table.TableSchema BQTableSchema) {
```

Comment:
Maybe better to name it as JsonWriterBuilder? So reader could easily understand what this Builder referring to?

Reply:
It's a static method so users will call it as JsonStreamWriter.newBuilder(), and it seems to be conventional to use newBuilder() (used in StreamWriter, Table.TableSchema, or when constructing protos) so I went with this. Should I still change the name?

Reply:
NVM.. looks good to me then